pkg/monitortests/kubeapiserver/apiservergracefulrestart/monitortest.go

```diff
@@ -0,0 +1,148 @@
+		//     for kube-apiserver, but it's not true for openshift-apiserver.
+		//	c) we expect both start and end shutdown events for a certain host
+		//     to appear and, do so in the desired order.
+		//  d) we do not expect nested shutdown window(s), for a certain host:
+		//      startW1, startW2, endW2, endW1
+		//  e) the apiserver creates the start and end shutdown events in order
+		//     with a significant delay in between, so it is highly unlikely that
+		//     the end event will be stored first in etcd, but it is possible that
+		//     creation of an event might fail and we see only one event from the pair.
+		// Links:
+		// a) https://github.com/openshift/kubernetes/blob/15f19ea2dd700767e5337502aec753d2a6e26905/staging/src/k8s.io/apiserver/pkg/server/config.go#L711-L742
+		// e) start: https://github.com/openshift/kubernetes/blob/15f19ea2dd700767e5337502aec753d2a6e26905/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go#L547
+		//    end: https://github.com/openshift/kubernetes/blob/15f19ea2dd700767e5337502aec753d2a6e26905/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go#L733
+		//
+		// One way to make this more deterministic would be to add a new carry in
+		// o/k with the following change:
+		//  - after we create the start event, save an object reference to this
+		//    start event just created.
+		//  - when we create the end event, set 'Related' field of the end event
+		//    to that of the start event.
+		// TODO: With the above approach, finding the pair will be more deterministic
+		podRef := monitorapi.PodFrom(currInterval.Locator)
+		nodeName, _ := monitorapi.NodeFromLocator(currInterval.Locator)
+		key := fmt.Sprintf("ns/%s pod/%s node/%s", podRef.Namespace, podRef.Name, nodeName)
+
+		switch reason {
+		case "ShutdownInitiated", "TerminationStart":
+			if _, ok := startedIntervals[key]; !ok {
+				startedIntervals[key] = currInterval.From
+			}
+		case "TerminationGracefulTerminationFinished":
+			startTime := beginning
+			if prevStart, ok := startedIntervals[key]; ok {
+				startTime = prevStart
+				delete(startedIntervals, key)
+			}
+
+			computedIntervals = append(computedIntervals,
+				monitorapi.NewInterval(monitorapi.SourceKubeEvent, monitorapi.Info).
+					Locator(monitorapi.NewLocator().
+						LocateServer(namespaceToServer[podRef.Namespace], nodeName, podRef.Namespace, podRef.Name, true),
+					).
+					Message(monitorapi.NewMessage().
+						Constructed("graceful-shutdown-analyzer").
+						Reason(monitorapi.GracefulAPIServerShutdown),
+					).
+					Display().
+					Build(startTime, currInterval.To),
+			)
+		}
+	}
+
+	// and now close everything still open with a warning
+	for fakeLocator, startTime := range startedIntervals {
+		podRef := monitorapi.PodFrom(fakeLocator)
+		nodeName, _ := monitorapi.NodeFromLocator(fakeLocator)
+
+		computedIntervals = append(computedIntervals,
+			monitorapi.NewInterval(monitorapi.SourceKubeEvent, monitorapi.Error).
```

Comment:
While I'm totally unsure if Source concept will survive once we start migrating the js UI to use these, I'd expect your Constructed("graceful-shutdown-analyzer") to instead be a Source here. I was thinking source is an indicator for what code produced the Interval and a tool to clearly identify what gets grouped into a section in the UI.

Reply:
Could you generally describe what a source is?  it is a MonitorTest?